ePBS specs (#1)
terencechain/consensus-specs: specs/_features/ePBS/validator.md

      - [Broadcast payload attestation](#broadcast-payload-attestation)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# ePBS -- Honest Validator

**Notice**: This document is a work-in-progress for researchers and implementers.

## Introduction

This document represents the changes to be made in the code of an "honest validator" to implement ePBS.

## Prerequisites

This document is an extension of the Deneb -- Honest Validator guide.
All behaviors and definitions defined in this document, and documents it extends, carry over unless explicitly noted or overridden.

All terminology, constants, functions, and protocol mechanics defined in the updated Beacon Chain doc of ePBS. are requisite for this document and used throughout.
Please see related Beacon Chain doc before continuing and use them as a reference throughout.

## Protocols

### `ExecutionEngine`

*Note*: `get_inclusion_list` function is added to the `ExecutionEngine` protocol for use as a validator.

The body of this function is implementation dependent.
The Engine API may be used to implement it with an external execution engine.

#### `get_inclusion_list`

> Given the `payload_id`, `get_payload` returns `GetInclusionListResponse` with the most recent version of

Reviewer: Are you sure you mean `get_payload`  here? Also what do you mean by " the execution payload that has been built since the last FCU call"?

`get_inclusion_list`  should simply return a current valid inclusion list.

> Given the `payload_id`, `get_payload` returns `GetInclusionListResponse` with the most recent version of

Reviewer: Why do we need a `payload_id`  here? More generally, we need to be explicit that calls to FCU should not be sending the payload attributes anymore since proposers are not builders. And in the honest builder spec this should be specified. It's important to note that the typical validator should not be triggering block creation on their EL when proposing.

Author: +1, I'll note this for the engine API spec. cc @naviechan

the execution payload that has been built since the corresponding call to `notify_forkchoice_updated` method.

```python

> def get_inclusion_list(self: ExecutionEngine, payload_id: PayloadId) -> GetInclusionListResponse:

Reviewer: The type ` GetInclusionListResponse`  needs to be defined

    """
    Return ``GetInclusionListResponse`` object.
    """
    ...
```

## Beacon chain responsibilities

All validator responsibilities remain unchanged other than those noted below. Namely, proposer normal block production switched to a new `BeaconBlockBody`.
Proposer with additional duty to construct and broadcast `InclusionList` alongside `SignedBeaconBlock`.
Attester by chance with an additional duty to be part of PTC committee and broadcast `PayloadAttestationMessage`.

## Validator assignment

> A validator can get PTC committee assignments for a given epoch using the following helper via `get_ptc_committee_assignment(state, epoch, validator_index)` where `epoch <= next_epoch`.

Reviewer: PTC assignments are by slot currently

A validator can use the following function to see if they are supposed to submit payload attestation message during a slot. 

> This function can only be run with a `state` of the slot in question. PTC committee selection is only stable within the context of the current and next epoch.

Reviewer: I think this should be a `state, slot`  pair and allow this function to be called with any state.

```python
def get_ptc_committee_assignment(state: BeaconState,
                             epoch: Epoch,
                             validator_index: ValidatorIndex
                             ) -> Optional[Tuple[Sequence[ValidatorIndex], Slot]]:
    """
    Return the ptc committee assignment in the ``epoch`` for ``validator_index``.
    ``assignment`` returned is a tuple of the following form:
        * ``assignment[0]`` is the list of validators in the committee
        * ``assignment[1]`` is the slot at which the committee is assigned
    Return None if no assignment.
    """
    next_epoch = Epoch(get_current_epoch(state) + 1)
    assert epoch <= next_epoch

    start_slot = compute_start_slot_at_epoch(epoch)
    committee_count_per_slot = get_committee_count_per_slot(state, epoch)
    for slot in range(start_slot, start_slot + SLOTS_PER_EPOCH):
        for index in range(committee_count_per_slot):

>             committee = get_ptc_committee(state, Slot(slot))

Reviewer: why aren't you using `get_ptc`  from the beacon-chain spec? https://github.com/potuz/consensus-specs/blob/epbs/specs/_features/epbs/beacon-chain.md#get_ptc

            if validator_index in committee:
                return committee, Slot(slot)
    return None
```

### Lookahead

The beacon chain shufflings are designed to provide a minimum of 1 epoch lookahead
on the validator's upcoming committee assignments for attesting dictated by the shuffling and slot.
Note that this lookahead does not apply to proposing, which must be checked during the epoch in question.

> `get_ptc_committee_assignment` should be called at the start of each epoch

Reviewer: a niptick: but ptc already has the word "committee"  in it

to get the assignment for the next epoch (`current_epoch + 1`).
A validator should plan for future assignments by noting their assigned ptc committee
slot and planning to participate in the ptc committee subnet.

Specifically a validator should:
* Find peers of the pubsub topic `beacon_attestation_{subnet_id}`.

>     * If the validator is assigned to be an aggregator for the slot (see `is_aggregator()`) [New in MaxEB], then subscribe to the topic.

Reviewer: `is_aggregator`  is not new in MaxEB, it's just modified.

### Inclusion list proposal

ePBS introduces forward inclusion list. The detail design is described in this [post](https://ethresear.ch/t/no-free-lunch-a-new-inclusion-list-design/16389)
Proposer is introduced to construct and broadcast `InclusionList` alongside `SignedBeaconBlock`.
- Proposer for slot `N` submits `SignedBeaconBlock` and in parallel submits `InclusionList` to be included at the beginning of slot `N+1`.
- Within `InclusionList`, `Transactions` are list of transactions that the proposer wants to include at the beginning of slot `N+1`.
- Within `inclusionList`, `Summaries` are lists consisting on addresses sending those transactions and their gas limits. The summaries are signed by the proposer.
- Proposer is allowed to send many of these pairs that aren't committed to its beacon block so no double proposing slashing is involved.

#### Constructing the `InclusionList`

To obtain an inclusion, a block proposer building a block on top of a `state` must take the following actions:

1. Retrieve inclusion list by calling `get_inclusion_list`.

```python
def get_inclusion_list(payload_id: Optional[PayloadId], execution_engine: ExecutionEngine) -> InclusionListResponse:

>         return execution_engine.get_payload(payload_id).inclusion_list

Reviewer: Why is there a call to `get_payload`  here? The inclusion list is proposed by the EL to be included in this slot. This function should have a signature equivalent to 
```get_inclusion_list(parent_block_hash: Bytes32) -> InclusionListResponse```

```

> 2. Build `InclusionListSummary`

Reviewer: I think the EL should return this and this should be part of the `InclusionListResponse`  class, the validator should simply sign it. Having the validator build the summary would imply the validator parsing the transactions slice, which we want to keep as opaque bytes in the CL

 - Set `summary` to the `inclusion_list_response.summary`
 - Set `proposer_index` to the proposer's index

3. Get `signature` by signing `InclusionListSummary` with its private key.

```python
def get_inclusion_list_summary_signature(state: BeaconState, inclusion_list_summary: InclusionListSummary, block_slot: Slot, privkey: int) -> BLSSignature:

>     domain = get_domain(state, DOMAIN_INCLUSION_LIST, compute_epoch_at_slot(block_slot))

Reviewer: Didn't we agree on using the Proposer domain here?

    return bls.Sign(privkey, signing_root)
```

4. Build `SignedInclusionListSummary`
 - set `signed_inclusion_list_summary.summary` to `InclusionListSummary`
 - set `signed_inclusion_list_summary.signature` to `signature

5. Build `InclusionList
- set `inclusion_list.summaries` to `SignedInclusionListSummary`
- set `inclusion_list.transactions` to `inclusion_list_response.transactions`

#### Broadcast inclusion list

Finally, the validator broadcasts inclusion to the inclusion list subnet, the `inclusion_list` pubsub topic.

### Block proposal

#### Constructing the new `signed_execution_payload_header_envelope` field in  `BeaconBlockBody`

To obtain `signed_execution_payload_header_envelope`, a block proposer building a block on top of a `state` must take the following actions:
* Listen to the `execution_payload_header` gossip subnet
* Filter out the bids where `signed_execution_payload_header_envelope.message.header.parent_hash` matches `state.latest_execution_payload_header.block_hash`
* The `signed_execution_payload_header_envelope` must satisfy the verification conditions found in `process_execution_payload_header`.

> * Select the best bid and set `body.signed_execution_payload_header_envelope = signed_execution_payload_header_envelope`

Reviewer: Do we want to specify "best bid"  here? I think proposers should be able to select "a bid".

#### Constructing the new `payload_attestations` field in  `BeaconBlockBody`

Up to `MAX_PAYLOAD_ATTESTATIONS`, aggregate payload attestations can be included in the block. 
The payload attestations added must satisfy the verification conditions found in payload attestation processing. 

> To maximize profit, the validator should attempt to gather aggregate payload attestations that include singular attestations from the largest number of validators whose signatures from the same epoch have not previously been added on chain.

Reviewer: `payload_attestations`  can only be included in the next slot, so there's only a maximum of 2 possible aggregates that are valid. We could modify `process_payload_attestations`  to allow for inclusion of previous ptc attestations in case of skipped slots or empty blocks, but complexity seems to be higher for no benefit.

### Attesting remains unchanged

### Attestation aggregation

Some validators are selected to locally aggregate attestations with a similar `attestation_data` to their constructed `attestation` for the assigned `slot`.

#### Aggregation selection

> A validator is selected to aggregate based upon the return value of `is_aggregator()`. [New in MaxEB]. Taken from [PR](https://github.com/michaelneuder/consensus-specs/pull/3)

Reviewer: ```suggestion
A validator is selected to aggregate based upon the return value of `is_aggregator()`. [Modified in ePBS]. Taken from [PR](https://github.com/michaelneuder/consensus-specs/pull/3)
```

```python
def is_aggregator(state: BeaconState, slot: Slot, index: CommitteeIndex, validator_index: ValidatorIndex, slot_signature: BLSSignature) -> bool:
    validator = state.validators[validator_index]
    committee = get_beacon_committee(state, slot, index)
    min_balance_increments = validator.effective_balance // MIN_ACTIVATION_BALANCE
    committee_balance_increments = get_total_balance(state, set(committee)) // MIN_ACTIVATION_BALANCE
    denominator = committee_balance_increments ** min_balance_increments
    numerator = denominator - (committee_balance_increments -  TARGET_AGGREGATORS_PER_COMMITTEE) ** min_balance_increments
    modulo = denominator // numerator
    return bytes_to_uint64(hash(slot_signature)[0:8]) % modulo == 0
```

Rest of the aggregation process remains unchanged.

### Payload timeliness attestation

> Some validators are selected to submit payload timeliness attestation. The `committee`, assigned `index`, and assigned `slot` for which the validator performs this role during an epoch are defined by `get_execution_committee(state, slot)`.

Reviewer: What is this function `get_execution_committee`?

A validator should create and broadcast the `execution_attestation` to the global execution attestation subnet at `SECONDS_PER_SLOT * 3 / INTERVALS_PER_SLOT` seconds after the start of `slot`

#### Constructing payload attestation

##### Prepare payload attestation message

> If a validator is in the payload attestation committee (i.e. `is_assigned_to_payload_committee()` above returns True), 

Reviewer: ```suggestion
If a validator is in the payload attestation committee (i.e. `is_assigned_to_payload_committee()` below returns True), 
```

then the validator should prepare a `PayloadAttestationMessage` for the current slot,
according to the logic in `get_payload_attestation_message` at `SECONDS_PER_SLOT * 3 / INTERVALS_PER_SLOT` interval.

```python

> def is_assigned_to_payload_committee(state: BeaconState,

Reviewer: Do we really need a helper for `validator_index in get_ptc(state, slot)`?

                                  validator_index: ValidatorIndex) -> bool:
                                                     
    committe = get_ptc(state, slot)
    return validator_index in committee
```

Next, the validator creates `payload_attestation_message` as follows:
* Set `payload_attestation_data.slot = slot` where `slot` is the assigned slot.
* Set `payload_attestation_data.beacon_block_root = block_root` where `block_root` is the block hash seen from the block builder reveal at `SECONDS_PER_SLOT * 2 / INTERVALS_PER_SLOT`. If the `SignedExecutionPayloadEnvelope`
* Set `payload_attestation_message.validator_index = validator_index` where `validator_index` is the validator chosen to submit. The private key mapping to `state.validators[validator_index].pubkey` is used to sign the payload timeliness attestation.
* Set `payload_attestation_message = PayloadAttestationMessage(data=payload_attestation_data, signature=payload_attestation_signature)`, where `payload_attestation_signature` is obtained from:

```python
def get_payload_attestation_message_signature(state: BeaconState, attestation: PayloadAttestationMessage, privkey: int) -> BLSSignature:

>     domain = get_domain(state, DOMAIN_PAYLOAD_TIMELINESS_COMMITTEE, compute_epoch_at_slot(attestation.slot))

Reviewer: in https://github.com/potuz/consensus-specs/blob/epbs/specs/_features/epbs/beacon-chain.md#domain-types there's `DOMAIN_PTC_ATTESTER`

    * If the validator is assigned to be an aggregator for the slot (see `is_aggregator()`)[Modified in MaxEB], then subscribe to the topic.

### Inclusion list proposal

ePBS introduces forward inclusion list. The detail design is described in this [post](https://ethresear.ch/t/no-free-lunch-a-new-inclusion-list-design/16389)
Proposer must construct and broadcast `InclusionList` alongside `SignedBeaconBlock`.
- Proposer for slot `N` submits `SignedBeaconBlock` and in parallel submits `InclusionList` to be included at the beginning of slot `N+1` builder.
- Within `InclusionList`, `Transactions` are list of transactions that the proposer wants to include at the beginning of slot `N+1` builder.
- Within `inclusionList`, `Summaries` are lists consisting on addresses sending those transactions and their gas limits. The summaries are signed by the proposer `N`.
- Proposer may  send many of these pairs that aren't committed to its beacon block so no double proposing slashing is involved.

#### Constructing the inclusion list

To obtain an inclusion list, a block proposer building a block on top of a `state` must take the following actions:

> 1. Retrieve inclusion list from execution layer by calling `get_execution_inclusion_list`.

Reviewer 2: I think proposer should also check if previous slot's payload was empty or not. If it was, the proposer shouldn't send a new inclusion list.

Reviewer 2: Okay I think this info has been added into the fork choice section by Potuz in a code snippet so not sure if we need it here.

## Prerequisites

This document is an extension of the Deneb -- Honest Validator guide.
All behaviors and definitions defined in this document, and documents it extends, carry over unless explicitly noted or overridden.

All terminology, constants, functions, and protocol mechanics defined in the updated Beacon Chain doc of ePBS. are requisite for this document and used throughout.
Please see related Beacon Chain doc before continuing and use them as a reference throughout.

## Beacon chain responsibilities

All validator responsibilities remain unchanged other than those noted below. Namely, proposer normal block production switched to using `SignedBeaconBlockWithBid` and attester with an additional duty to submit `ExecutionAttestation`

### Block proposal

#### Constructing the `BeaconBlockBody`

To obtain an signed builder bid, a block proposer building a block on top of a `state` must take the following actions:
* Listen to the `builder_bid` gossip subnet
* Filter the bids where where `bid.header.parent_hash` matches `head.payload.block_hash`
* Select the best bid (the one with the highest value) and set `body.bid = signed_builder_bid`

> ### Consensus attesting reminds unchanged

Reviewer 3: ```suggestion
### Consensus attesting remains unchanged
```

### Consensus attestation aggregation reminds unchanged

### Payload timeliness attestation

Some validators are selected to submit payload timeliness attestation. The `committee`, assigned `index`, and assigned `slot` for which the validator performs this role during an epoch are defined by `get_execution_committee(state, slot)`.

A validator should create and broadcast the `execution_attestation` to the global execution attestation subnet at `SECONDS_PER_SLOT * 3 / INTERVALS_PER_SLOT` seconds after the start of `slot`

#### Construct payload timeliness attestation

Next, the validator creates `signed_execution_attestation`
* Set `execution_attestation.slot = slot` where `slot` is the assigned slot.

> * Set `execution_attestation.block_hash = block_hash` where `block_hash` is the block hash seen from the block builder reveal at `SECONDS_PER_SLOT * 2 / INTERVALS_PER_SLOT`

Reviewer 4: Here block_hash refers to the `SignedBeaconBlockWithBid` hash tree root which is head at the builder reveal deadline?

* Set `execution_attestation.validator_index = validator_index` where `validator_index` is the validator chosen to submit. The private key mapping to `state.validators[validator_index].pubkey` is used to sign the payload timeliness attestation.

> * Set `signed_execution_attestation = SignedExecution(message=execution_attestation, signature=execution_attestation_signature)`, where `execution_attestation_signature` is obtained from:

Reviewer 4: The PTC should cast a vote for either full or empty, does this data-structure allow empty votes?

Author: Perhaps empty could be leveraged as zero hashes?

Reviewer 4: You still need to indicate that the payload for specifically header block_hash was not seen so that field must be set
